File: plus/python/src/planner/tax/compensacion.py
"""Compensación de rentas negativas with carry-forward to later years (arts. 49 and 50.3 LIRPF).

The carry-forward state is an array of loss amounts (positive numbers) by year of origin,
oldest first; the last axis has length `anos` (4). For the savings base there is one more axis
before it with two baskets: [RCM — rendimientos del capital mobiliario, GP — ganancias y pérdidas].
Functions return the new state for the next year: the oldest year drops out, the current year's
unused losses are appended.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from planner.tax.rules import Compensacion

RCM, GP = 0, 1


def _out(x):
    return float(x) if np.ndim(x) == 0 else x


def pendientes_vacios(rules: Compensacion, shape: tuple[int, ...] = (), ahorro: bool = True):
    """Empty carry-forward state: (..., 2, anos) for ahorro or (..., anos) for general."""
    return np.zeros((*shape, 2, rules.anos) if ahorro else (*shape, rules.anos))


def _consumir(importe, pendientes):
    """Consumes up to importe from pendientes (oldest first); returns the amount used per basket."""
    importe = np.asarray(importe, dtype=float)[..., None]
    antes = np.cumsum(pendientes, axis=-1) - pendientes
    return np.clip(importe - antes, 0.0, pendientes)


def _avanzar(pendientes, nuevas):
    """Shifts one year: the oldest year drops out, the current year's losses go to the end."""
    return np.concatenate([pendientes[..., 1:], np.asarray(nuevas)[..., None]], axis=-1)


@dataclass(frozen=True, slots=True)
class BaseAhorro:
    base_imponible: float
    compensado_anteriores: float  # prior-year losses used this year
    pendientes: np.ndarray  # state for the next year, (..., 2, anos)


def base_imponible_ahorro(rcm, ganancias, pendientes, rules: Compensacion) -> BaseAhorro:
    """Base imponible del ahorro for the year with loss offsetting (art. 49; order per the
    Manual práctico).

    rcm — saldo of rendimientos del capital mobiliario (interest, dividends); may be < 0.
    ganancias — saldo of ganancias y pérdidas patrimoniales for the year; may be < 0.

    Step 1: a negative saldo in one basket is offset by the other basket's positive saldo, but by
    no more than limite_cruzado of it. Step 2: prior-year losses are offset against their own
    basket in full first, then against the other within what is left of the same limite_cruzado
    (shared with step 1).
    """
    rcm = np.asarray(rcm, dtype=float)
    gp = np.asarray(ganancias, dtype=float)
    pend = np.array(pendientes, dtype=float, copy=True)
    lim = rules.limite_cruzado

    disponible = {RCM: np.maximum(rcm, 0.0), GP: np.maximum(gp, 0.0)}
    # How much of the other basket's losses can be offset here (25 % of its positive saldo).
    cupo = {k: lim * v for k, v in disponible.items()}
    otra = {RCM: GP, GP: RCM}

    # Step 1: current-year losses are offset by the other basket within cupo.
    nuevas = {}
    for k, saldo in ((RCM, rcm), (GP, gp)):
        perdida = np.maximum(-saldo, 0.0)
        c = np.minimum(perdida, cupo[otra[k]])
        disponible[otra[k]] = disponible[otra[k]] - c
        cupo[otra[k]] = cupo[otra[k]] - c
        nuevas[k] = perdida - c

    def compensar(k_perdida: int, k_saldo: int, tope) -> np.ndarray:
        usado = _consumir(tope, pend[..., k_perdida, :])
        pend[..., k_perdida, :] -= usado
        s = usado.sum(axis=-1)
        disponible[k_saldo] = disponible[k_saldo] - s
        return s

    # Step 2.1: prior-year losses — against their own basket without limit.
    total = compensar(RCM, RCM, disponible[RCM]) + compensar(GP, GP, disponible[GP])
    # Step 2.2: the rest — against the other basket within the remaining cupo.
    total = total + compensar(RCM, GP, np.minimum(disponible[GP], cupo[GP]))
    total = total + compensar(GP, RCM, np.minimum(disponible[RCM], cupo[RCM]))

    shape = np.broadcast_shapes(rcm.shape, gp.shape)
    nuevas_arr = np.stack([np.broadcast_to(nuevas[k], shape) for k in (RCM, GP)], axis=-1)
    return BaseAhorro(
        base_imponible=_out(disponible[RCM] + disponible[GP]),
        compensado_anteriores=_out(total),
        pendientes=_avanzar(pend, nuevas_arr),
    )


@dataclass(frozen=True, slots=True)
class BaseGeneral:
    base_liquidable: float  # after offsetting prior negative bases, ≥ 0
    compensado_anteriores: float
    pendientes: np.ndarray  # (..., anos)


def compensar_base_liquidable_general(base_liquidable, pendientes, rules: Compensacion):
    """A negative base liquidable general is carried forward 4 years (art. 50.3 LIRPF)."""
    blg = np.asarray(base_liquidable, dtype=float)
    pend = np.asarray(pendientes, dtype=float)
    positiva = np.maximum(blg, 0.0)
    usado = _consumir(positiva, pend)
    s = usado.sum(axis=-1)
    return BaseGeneral(
        base_liquidable=_out(positiva - s),
        compensado_anteriores=_out(s),
        pendientes=_avanzar(pend - usado, np.maximum(-blg, 0.0)),
    )

File: plus/addon/src/es-tax/compensacion.ts
// Порт planner.tax.compensacion: зачёт убытков с переносом на 4 года (arts. 49, 50.3 LIRPF).
// Состояние — суммы убытков по годам происхождения, от старшего к младшему. Для базы
// сбережений — две корзины: [RCM, GP].
import type { Compensacion } from './rules';

export type PendientesGeneral = number[];
export type PendientesAhorro = [number[], number[]];

const RCM = 0;
const GP = 1;

export function pendientesVaciosGeneral(rules: Compensacion): PendientesGeneral {
  return new Array(rules.anos).fill(0);
}

export function pendientesVaciosAhorro(rules: Compensacion): PendientesAhorro {
  return [new Array(rules.anos).fill(0), new Array(rules.anos).fill(0)];
}

/** Списать до importe из корзин (старшие первыми); списанное по корзинам. */
function consumir(importe: number, pendientes: number[]): number[] {
  let antes = 0;
  return pendientes.map((p) => {
    const usado = Math.min(Math.max(importe - antes, 0), p);
    antes += p;
    return usado;
  });
}

/** Сдвиг на год: старший выбывает, в конец — убытки текущего года. */
const avanzar = (pendientes: number[], nuevas: number) => [...pendientes.slice(1), nuevas];

const sum = (xs: number[]) => xs.reduce((a, b) => a + b, 0);

export interface BaseAhorro {
  base_imponible: number;
  compensado_anteriores: number;
  pendientes: PendientesAhorro;
}

/**
 * Base imponible del ahorro с зачётом убытков. Фаза 1: убыток одной корзины гасится
 * положительным saldo другой в пределах limite_cruzado. Фаза 2: прошлые убытки — своей
 * корзиной целиком, затем другой в пределах остатка того же лимита.
 */
export function baseImponibleAhorro(
  rcm: number,
  ganancias: number,
  pendientes: PendientesAhorro,
  rules: Compensacion,
): BaseAhorro {
  const pend = pendientes.map((p) => [...p]) as PendientesAhorro;
  const lim = rules.limite_cruzado;
  const saldo = [rcm, ganancias];
  const disponible = [Math.max(rcm, 0), Math.max(ganancias, 0)];
  const cupo = disponible.map((v) => lim * v);
  const otra = (k: number) => 1 - k;

  const nuevas = [0, 0];
  for (const k of [RCM, GP]) {
    const perdida = Math.max(-saldo[k], 0);
    const c = Math.min(perdida, cupo[otra(k)]);
    disponible[otra(k)] -= c;
    cupo[otra(k)] -= c;
    nuevas[k] = perdida - c;
  }

  const compensar = (kPerdida: number, kSaldo: number, tope: number) => {
    const usado = consumir(tope, pend[kPerdida]);
    pend[kPerdida] = pend[kPerdida].map((p, i) => p - usado[i]);
    const s = sum(usado);
    disponible[kSaldo] -= s;
    return s;
  };

  let total = compensar(RCM, RCM, disponible[RCM]);
  total += compensar(GP, GP, disponible[GP]);
  total += compensar(RCM, GP, Math.min(disponible[GP], cupo[GP]));
  total += compensar(GP, RCM, Math.min(disponible[RCM], cupo[RCM]));

  return {
    base_imponible: disponible[RCM] + disponible[GP],
    compensado_anteriores: total,
    pendientes: [avanzar(pend[RCM], nuevas[RCM]), avanzar(pend[GP], nuevas[GP])],
  };
}

export interface BaseGeneral {
  base_liquidable: number; // ≥ 0
  compensado_anteriores: number;
  pendientes: PendientesGeneral;
}

/** Отрицательная base liquidable general переносится на 4 года (art. 50.3). */
export function compensarBaseLiquidableGeneral(
  baseLiquidable: number,
  pendientes: PendientesGeneral,
): BaseGeneral {
  const positiva = Math.max(baseLiquidable, 0);
  const usado = consumir(positiva, pendientes);
  const s = sum(usado);
  return {
    base_liquidable: positiva - s,
    compensado_anteriores: s,
    pendientes: avanzar(
      pendientes.map((p, i) => p - usado[i]),
      Math.max(-baseLiquidable, 0),
    ),
  };
}
